pennylane/ftqc/qubit_graph.py

```diff
@@ -0,0 +1,601 @@
+            >>> print(f"{q_top}; {q_top.nodes}")
+            QubitGraph<top>; [0]
+            >>> q_new = QubitGraph("new", graph)
+            >>> print(f"{q_new}; {q_new.nodes}")
+            QubitGraph<new>; [0]
+            >>> q_top[0] = q_new
+            >>> print(f"{q_top[0]}; {q_top[0].nodes}")
+            QubitGraph<top, 0>; [0]
+
+        ..  TODO:
+
+            - Allow for more advanced tensor-like indexing and slicing.
+            - Explicitly disallow assigning a QubitGraph object that would make the nesting
+              structure cyclic.
+        """
+        if not isinstance(value, QubitGraph):
+            raise TypeError(
+                f"'QubitGraph' item assignment type must also be a QubitGraph, but got type "
+                f"'{type(value).__name__}'"
+            )
+
+        if not self.is_initialized:
+            self._warn_uninitialized()
+            return
+
+        value._id = key
+        value._parent = self
+
+        self._graph_qubits.nodes[key]["qubits"] = value
+
+    def __iter__(self):
+        """Dummy QubitGraph iterator method that yields itself.
+
+        This method ensures that no implicit iteration takes place over the nodes in the underlying
+        qubit graph, which is possible since we have defined ``__getitem__()``.
+
+        Consider the example where we wrap a QubitGraph object in a tuple. With this dummy
+        ``__iter__()`` method defined, the resulting tuple contains the top-level QubitGraph object:
+
+            >>> q = QubitGraph(0)
+            >>> q.init_graph_nd_grid((2,))
+            >>> tuple(q)
+            (QubitGraph<0>,)
+
+        Without this dummy method defined, the ``tuple()`` constructor attempts to iterate over the
+        values that the ``__getitem__()`` method yield, which creates a tuple of the QubitGraph
+        objects contained in the underlying qubit graph:
+
+            >>> # Without QubitGraph.__iter__() defined
+            >>> q = QubitGraph(0)
+            >>> q.init_graph_nd_grid((2,))
+            >>> tuple(q)
+            (QubitGraph<0, 0>, QubitGraph<0, 1>)
+
+        Making QubitGraph non-iterable is especially important when using it as input to the
+        :class:`~.Wires` class, which checks if the input is iterable by wrapping it in a tuple.
+        """
+        yield self
+
+    def __repr__(self) -> str:
+        """Representation of a QubitGraph object.
+
+        This representation displays the full index hierarchy of a nested QubitGraph object.
+
+        Examples:
+
+            >>> QubitGraph(0)
+            QubitGraph<0>
+
+            >>> q = QubitGraph(0)
+            >>> q.init_graph_nd_grid((2, 2))
+            >>> q
+            QubitGraph<0>
+            >>> q[(0, 1)]
+            QubitGraph<0, (0, 1)>
+            >>> graph = nx.Graph()
+            >>> graph.add_node("aux")
+            >>> q[(0, 1)].init_graph(graph)
+            >>> q[(0, 1)]["aux"]
+            QubitGraph<0, (0, 1), aux>
```

Comment:
When I was playing around with this, I found that it would help me if the format were something like:

```suggestion
            QubitGraph<id=0, node_location=[(0, 1), aux]>
```

I don't know if the name `node_location` is a very good idea, its just the first thing that came to mind, but some way of separating these to make it clearer when I want to index into the object, the very first thing is _not_ an index, but the overall "name".

Reply:
Yes I think we can improve this representation. For example, in this case the ID of the QubitGraph at `q[(0, 1)]["aux"]` is actually "aux":

```pycon
>>> q = QubitGraph(0)
>>> q.init_graph_nd_grid((2, 2))
>>> graph = nx.Graph()
>>> graph.add_node("aux")
>>> q[(0, 1)].init_graph(graph)
>>> q[(0, 1)]["aux"].id
'aux'
```

So if that wasn't obvious we should definitely change it! :sweat_smile:

Maybe this would be a better representation:

```pycon
q
QubitGraph<id=0, loc=root>
```

which would be read as "this is QubitGraph 0 and it is a root node". And then:

```pycon
>>> q[(0, 1)]["aux"]
QubitGraph<id='aux', loc=[(0, 1), 0]>
```

which would be read as "this is QubitGraph 'aux', whose parent is the node with label '(0, 1)', whose parent is node with label '0', which is a root node."

Reply:
Done. 03edcb84a